### ISSUE TITLE
Improving testing for ANN layers

### ISSUE THREAD

**Issue by mrdaybird** [mlpack/mlpack#3472]
**Summary** 

- The current testing methodology for testing ANN Layers is inadequate and unreliable,(evident with issues such as #3468, and #3478 )
-  There should be certain checklists for testing **correctness** and **reliability** of a certain layer before it can be used as part of mlpack. 

**Motivation** 

1. Many tests for layers improperly call the _Backward_ method, such as AddMerge, LogSoftMax etc. 
2. There is no verification of the external data used in tests.
3. If external data is used in tests, there are no guidelines for correct sources of data or how this data should be created.
4. Most tests do not use JacobianTest for testing the correctness of derivative, which maybe a reliable test for correctness of derivative as it does not use external data. Also, JacobianTest up till now also incorrectly used _Backward_ method, which may be a possible reason for it's lack of use in testing. (See #3470)
5. Some tests do not test their _Backward_ method. Such as BatchNorm layer. 
6. Some tests use trivial example for testing the Backward method, which does not verify correctness of Backward. Such as Concat which tests against zero matrix. 

**Ideas**

1. All layers test should use JacobianTest for testing correctness of derivative.
2. Layers should be tested in combination with other layers, in form of a neural network that tests whether the neural network is able to learn anything at all. 
      1. For Activation function, a template model, in which the activation function is our template, could be written that trains and infers against the same dataset. Finally the accuracy can be compared with respect to the activation functions.

If you have any other ideas or thoughts please share.

**Current work**

PRs which aims to solve the mentioned issues:
1. #3478 documents the correct usage of Backward/Deriv, it also adds JacobianTest for all layers and fix several issues.

**Comment by shubham1206agra** [comment]
https://colab.research.google.com/drive/18wRACRRgt6Q97dd7SWjadWofHyD8wDnV?usp=sharing

**Comment by gullayeshwantkumarruler** [comment]
Hi Shubham, Yes, I too think that the ANN layers are not reliable so it would be good to create some checklists for testing the reliability and correctness.
